src/compiler/compiler.rs

```diff
@@ -151,25 +173,31 @@ impl<F: Field + Hash> Compiler<F> {
                 self.add_step_type_handlers(ctx, symbols, machine_id);
 
                 for state_id in machine.keys() {
-                    ctx.step_type_def(self.get_step_type_handler(machine_id, state_id), |ctx| {
-                        self.add_internals(ctx, symbols, machine_id, state_id);
-
-                        ctx.setup(|ctx| {
-                            let pis = self.translate_queries(symbols, setup, machine_id, state_id);
-                            pis.iter().for_each(|pi| {
-                                let constraint = Constraint {
-                                    annotation: format!("{:?}", pi),
-                                    expr: pi.clone(),
-                                    typing: Typing::AntiBooly,
-                                };
-                                ctx.constr(constraint);
+                    ctx.step_type_def(
+                        self.mapping.get_step_type_handler(machine_id, state_id),
+                        |ctx| {
+                            self.add_internals(ctx, symbols, machine_id, state_id);
+
+                            ctx.setup(|ctx| {
+                                let pis =
+                                    self.translate_queries(symbols, setup, machine_id, state_id);
+                                pis.iter().for_each(|pi| {
+                                    let constraint = Constraint {
+                                        annotation: format!("{:?}", pi),
+                                        expr: pi.clone(),
+                                        typing: Typing::AntiBooly,
```

Comment:
AntiNooly is a funny name :) Does it mean a negated boolean?

Reply:
Booly is boolean like in C, 0 is false anything else is true. Anti-booly is the oposite, 0 is true, anything else is false. Then we have OneZero (1 true, 0 false, it is not possible something else).

To convert arbitrary boolean expressions into PI we need to use both represenations and do casting between them. Casting OneZero to Anti-Booly is 1-x; The other way around is the is_zero gadget

Reply:
It would be cool to have it as a doc above the type enum.

Reply:
there is this, but I think there are some mistakes in the code

https://github.com/leolara/notes/blob/main/drafts/eugenio.md